lib/realtime/data_status.ex

```diff
@@ -1,8 +1,51 @@
 defmodule Realtime.DataStatus do
+  alias Realtime.Vehicle
+
   @type t :: :good | :outage
 
-  @spec data_status() :: t()
-  def data_status() do
-    :good
+  @spec data_status([Vehicle.t()]) :: t()
+  def data_status(vehicles) do
+    now_fn = Application.get_env(:skate, :now_fn, &Util.Time.now/0)
+    now = now_fn.()
+
+    {good_count, bad_count} =
+      vehicles
+      |> Enum.filter(&should_consider?/1)
+      |> count_partition(fn vehicle -> vehicle_is_good(vehicle, now) end)
+
+    total_count = good_count + bad_count
+
+    if bad_count >= 2 and bad_count / total_count > 0.20 do
+      :outage
+    else
+      :good
+    end
+  end
+
+  # Our data often has lots of out of service buses that have
+  # good reasons for going stale.
+  # Only consider buses that we expect and need to have good data
+  @spec should_consider?(Vehicle.t()) :: boolean()
+  defp should_consider?(vehicle) do
+    vehicle.route_id != nil
+  end
+
+  @spec vehicle_is_good(Vehicle.t(), Util.Time.timestamp()) :: boolean()
+  defp vehicle_is_good(vehicle, now) do
+    age_seconds = now - vehicle.timestamp
+    # 2 and a half minutes ago.
+    age_seconds < 150
+  end
+
+  @spec count_partition([element], (element -> bool())) :: {non_neg_integer(), non_neg_integer()}
```

Comment:
You don't actually care about the number of good vehicles, so you could have this return `{predicate_count, total_count}` and change the predicate to `vehicle_is_bad`.

Reply:
I would go further and say that I'm not sure you need to partition at all. You already know the total count: it is `vehicles |> Enum.filter(&should_consider?/1) |> length()`. You also know the threshold of bad results you have to cross, the max of `count * 0.2` and `3`, so you don't even actually care about the bad count. I think you could probably use `Stream.take_while/2` or something and lazily evaluate such that you can stop as soon as you've crossed the threshold.

Reply:
`length()` is linear time, and we at least need that, so this needs to take linear time no matter what. Using `take_while` for the second step doesn't save us from that.

As for getting the total count, I chose this way to stay the same as [`Enum.partition()`](https://hexdocs.pm/elixir/1.0.5/Enum.html#partition/2), which returns `{true_results, false_results}`.